doc/custom-widgets.md

```diff
@@ -0,0 +1,139 @@
+## Customize widgets
+
+Angular InstantSearch comes with a set of UI widgets. Some of them have options to modify the HTML output (`<ng-template>`).
+
+Every of these widgets are built using the [InstantSearch.js connectors API](https://community.algolia.com/instantsearch.js/v2/connectors.html) which contains the logic of the widgets without the rendering.
+
+### Introduce BaseWidget class
+
+The `BaseWidget` class helps you to create new widgets using the InstantSearch.js connectors with Angular.
+
+It encapsulate the logic to maintain the state of the widget in sync with the search and handles for you the initialisation and the dispose of the widget on Angular life cycle hooks.
+
+Let's go step by step on how to write a custom widget using the `BaseWidget` class.
+
+## Custom Menu Select widget
+
+In this example we will create a new custom widget using the `BaseWidget` class and the [connectMenu](https://community.algolia.com/instantsearch.js/v2/connectors/connectMenu.html) InstantSearch.js connector.
+
+### 1. Extending the `BaseWidget` class
+
+First of all, you will need to write some boilerplate code in order to initialize correctly the `BaseWidget` class. This happens in the `constructor()` of your class extending the `BaseWidget` class.
+
+```ts
+import { Component, Inject, PLATFORM_ID } from '@angular/core';
+import { BaseWidget, NgAisInstance } from 'angular-instantsearch';
+
+@Component({
+  selector: 'ng-ais-menu-select',
+  template: '<p>It works!</p>'
+})
+export class MenuSelect extends BaseWidget {
+  constructor(
+    @Inject(PLATFORM_ID) public platformId: Object,
+    searchInstance: NgAisInstance
+  ) {
+    super(searchInstance, 'MenuSelect');
+  }
+}
+```
+
+* We need to indicate the `BaseWidget` class if we are in a browser or server environment
+* We need to indicate also the widget name in order to generate the CSS classes
+
+We have the first code needed for our MenuSelect widget, let's move on with the connector part!
+
+### 2. Inject the `connnectMenu` connector
+
+The `BaseWidget` class has a method called `createWidget()` which takes at first parameter the connector to use and as second parameters an object of options for this connector.
+
+To know the options used by the connector we can refer to the [InstantSearch.js connectors API documentation](https://community.algolia.com/instantsearch.js/v2/connectors.html).
+
+In our case, we will use the `connectMenu` which accepts multiple options but for simplicity we will only use the `attributeName` option which is the only one mandatory.
+
+(The `attributeName` is the name of the attribute for faceting)
+
+```ts
+import { Component, Inject, PLATFORM_ID } from '@angular/core';
+import { BaseWidget, NgAisInstance } from 'angular-instantsearch';
+import { connectMenu } from 'instantsearch.js/es/connectors';
```

Comment:
As soon as we do this, it means we are introducing a new public API (connectors). I am fine with it but now we will have to be extra careful with it (and also at some point be able to remove that /es/), @samouss I remember a previous discussion on how to remove this /es/, am I right? Any pointers?

Reply:
The `/es/` is needed for having a complete tree-shaking and the possibility to target the exact files.

We document the `/es/` prefix on IS.js here: https://community.algolia.com/instantsearch.js/v2/guides/usage.html#optimize-your-build-with-tree-shaking